js/web/lib/wasm/jsep/webgpu/ops/argminmax.ts

```diff
@@ -0,0 +1,156 @@
+
+          ${idxCopy.join('\n')}
+          ${ops[0]}       // init ops
+          ${initInputIdx}
+          ${ops[1]}
+          ${reduceOps}
+          ${ops[3]} // final values
+          output[global_idx*2] = bestIndex; // result it int64
+        }`;
+
+      return {
+        ...metadata,
+        getShaderSource,
+        outputs: [{dims: outputShape, dataType: inputs[0].dataType, gpuDataType: GpuDataType.default}],
+        dispatchGroup: () => ({x: Math.ceil(outputSize / 64)})
+      };
+    };
+
+const createArgMinMaxAttributesFromInputs =
+    (inputs: readonly TensorView[], attributes: ArgMinMaxAttributes): ArgMinMaxAttributes =>
+        createAttributeWithCacheKey(
+            {axes: attributes.axes, keepDims: attributes.keepDims, selectLastIndex: attributes.selectLastIndex});
+
+const createReduceProgramInfoLoader =
+    (inputs: readonly TensorView[], name: string, attributes: ArgMinMaxAttributes, reduceOp: ArgMinMaxOp):
+        ProgramInfoLoader => {
+          const updatedAttributes: ArgMinMaxAttributes =
+              inputs.length === 1 ? attributes : createArgMinMaxAttributesFromInputs(inputs, attributes);
+          const metadata:
+              ProgramMetadata = {name, inputTypes: [GpuDataType.default], cacheHint: updatedAttributes.cacheKey};
+          return {...metadata, get: () => createReduceProgramInfo(metadata, [inputs[0]], updatedAttributes, reduceOp)};
+        };
+
+
+export const argMin = (context: ComputeContext, attributes: ArgMinMaxAttributes): void => {
+  validateInputs(context.inputs);
+  const argMinMaxOp: ArgMinMaxOp = (inputs: TensorView[], axes: number[]): string[] => {
+    const idxZero = [];
+    for (let k = 0; k < inputs[0].dims.length; k++) {
+      if (axes.indexOf(k) >= 0 || axes.length === 0) {
+        idxZero.push(`inputIndices[${k}] = 0;`);  // first element
+      }
+    }
+    return [
+      `${idxZero.join('\n')}`, 'var value = _A[inputIdx];\nvar bestIndex : i32 = 0;',
+      'if (_A[inputIdx] < value) {value = _A[inputIdx]; bestIndex = i32(lastIndex);} ', ''
+    ];
+  };
+  context.compute(createReduceProgramInfoLoader(context.inputs, 'ArgMin', attributes, argMinMaxOp), {inputs: [0]});
+};
+
+export const argMax = (context: ComputeContext, attributes: ArgMinMaxAttributes): void => {
+  validateInputs(context.inputs);
+  const argMinMaxOp: ArgMinMaxOp = (inputs: TensorView[], axes: number[]): string[] => {
+    const idxZero = [];
+    for (let k = 0; k < inputs[0].dims.length; k++) {
+      if (axes.indexOf(k) >= 0 || axes.length === 0) {
+        idxZero.push(`inputIndices[${k}] = 0;`);  // first element
+      }
+    }
+    return [
+      `${idxZero.join('\n')}`, 'var value = _A[inputIdx];\nvar bestIndex : i32 = 0;',
+      'if (_A[inputIdx] > value) {value = _A[inputIdx]; bestIndex = i32(lastIndex);} ', ''
+    ];
+  };
+  context.compute(createReduceProgramInfoLoader(context.inputs, 'argMax', attributes, argMinMaxOp), {inputs: [0]});
```

Comment:
Would it easy to cast the output from the compute call from 32-bit integer to 64-bit integer rather writing 32-bit results into 64-bit results?

Reply:
I did not consider it for 2 reasons:
1. inside wgsl there is no int64 type so I would need to use a onnx cast on cpu which would be very costly
2. there is no onnx signature for argmin/argmax that has a i32
For now this route seems to be the easiest. Might need to revisit once we have other ops that deal with int64.